### DISCUSSION TITLE
Let's talk about v4.x

### DISCUSSION THREAD

**Discussion by VonHeikemen** [VonHeikemen/lsp-zero.nvim#402] [Ideas]
The [branch for v4.x](https://github.com/VonHeikemen/lsp-zero.nvim/tree/v4.x) has been created. If everyone is ok with it I could make the default in `august`, or is that too soon? The code is already written, I'm just missing the documentation.

Let me know what you think about these changes.

## New getting started section

In the "getting started" section I'll try my best to tell the reader `nvim-lspconfig` is the plugin they are using to configure the language server.

The result from the getting started should be something like this for the LSP configuration.

```lua
local lsp_zero = require('lsp-zero')

local lsp_attach = function(client, bufnr)
  local opts = {buffer = bufnr}

  vim.keymap.set('n', 'K', '<cmd>lua vim.lsp.buf.hover()<cr>', opts)
  vim.keymap.set('n', 'gd', '<cmd>lua vim.lsp.buf.definition()<cr>', opts)
  vim.keymap.set('n', 'gD', '<cmd>lua vim.lsp.buf.declaration()<cr>', opts)
  vim.keymap.set('n', 'gi', '<cmd>lua vim.lsp.buf.implementation()<cr>', opts)
  vim.keymap.set('n', 'go', '<cmd>lua vim.lsp.buf.type_definition()<cr>', opts)
  vim.keymap.set('n', 'gr', '<cmd>lua vim.lsp.buf.references()<cr>', opts)
  vim.keymap.set('n', 'gs', '<cmd>lua vim.lsp.buf.signature_help()<cr>', opts)
  vim.keymap.set('n', '<F2>', '<cmd>lua vim.lsp.buf.rename()<cr>', opts)
  vim.keymap.set({'n', 'x'}, '<F3>', '<cmd>lua vim.lsp.buf.format({async = true})<cr>', opts)
  vim.keymap.set('n', '<F4>', '<cmd>lua vim.lsp.buf.code_action()<cr>', opts)
end

lsp_zero.extend_lspconfig({
  sign_text = true,
  lsp_attach = lsp_attach,
  capabilities = require('cmp_nvim_lsp').default_capabilities(),
})

-- These are just examples. Replace them with the language
-- servers you have installed in your system
require('lspconfig').gleam.setup({})
require('lspconfig').rust_analyzer.setup({})
```

And the autocompletion setup is just the minimal needed to get nvim-cmp working.

```lua
local cmp = require('cmp')

cmp.setup({
  sources = {
    {name = 'nvim_lsp'},
  },
  snippet = {
    expand = function(args)
      -- You need Neovim v0.10 to use vim.snippet
      vim.snippet.expand(args.body)
    end,
  },
  mapping = cmp.mapping.preset.insert({}),
})
```

I would love to hide a lot of this nonsense and make it automatic, the problem is users asking for help on other online platforms don't get the help they need. Other Neovim users who can help don't know they can help. If all they see is `lsp_zero.something`, they say "I don't know that plugin. Can't help you." So, you know, I give up on the idea of adding automatic magic. Configuration will be explicit and verbose.

### `v4.x` will drop the automatic setup of nvim-lspconfig

The user will now have to call `.extend_lspconfig()` function and provide the `cmp_nvim_lsp` capabilities settings.

### `v4.x` will drop the automatic setup of nvim-cmp

Neovim v0.10 has its own snippet engine, and nvim-cmp tries to use it (if available).

The function `.extend_cmp()` still works and it does the same thing as the example in the getting started section.

### Deprecated functions were removed

These function were marked as deprecated in `v2.x`. In version `v3.x` they were empty or showed a warning message. Now they are really gone.

* `.preset()`
* `.ensure_installed()`
* `.setup_nvim_cmp()`
* `.skip_server_setup()`
* `.nvim_workspace()`
* `.set_preferences()`
* `.defaults.cmp_config()`
* `.defaults.cmp_mappings()`

### `.build_options()` function was removed

People even use it? I think I just forgot to delete it in `v3.x`.

## Migrating from v3.x to v4.x

The API is still 99% compatible. Changing from one branch to another should be painless... unless of course one of the deprecated functions is being used. Those are safe to remove even in `v3.x`, they have no effect.

On the LSP side users have to opt-in to the features that were previously enabled by default.

* Reserve the space in the signcolumn.
* Configure the border style of floating windows
* Provide the capabilities to `.extend_lspconfig()`

```lua
local lsp_zero = require('lsp-zero')

lsp_zero.extend_lspconfig({
  sign_text = true,
  float_border = 'rounded',
  capabilities = require('cmp_nvim_lsp').default_capabilities(),
})
```

For autocompletion, users would have to call `.extend_cmp()` 

```lua
local lsp_zero = require('lsp-zero')

lsp_zero.extend_cmp()
```

or provide the basic options to nvim-cmp's setup.

```lua
local cmp = require('cmp')

cmp.setup({
  sources = {
    {name = 'nvim_lsp'},
  },
  mapping = cmp.mapping.preset.insert({}),
})
```

**Comment by pm4rcin** [comment]
Since https://github.com/folke/lazy.nvim/issues/338 is now implemented I think it would be beneficial to configure Laziness in this repo so it would be one less thing to do for users. What do you think?

**Reply by VonHeikemen** [reply]
Laziness as in you only execute the code you need is already in the source code of the plugin.

In my own setup, which uses a fair amount of features, `lazy.nvim` reports lsp-zero takes `0.4ms` to load. And nvim-lspconfig takes `2.34ms` to load. I honestly think those numbers are pretty good, and I use a very old desktop PC from 2011.

**Comment by boydkelly** [comment]
Hi,  Please bear in mind this is from a user perspetive.   I have (by necessity) had to use Helix and Zed to edit my yaml files.   lsp just works once you download the tools.   That said...  I believe the future of neovim will/should be one day....  (And can it be v4.x?)

opts = {
ensure_insalled = {...}
auto_update = true
...
}
require ("lspzero").setup({opts})

No mason, mason-lspconfig....   Just pair up with mason-tool-installer to get the tools and a one-liner to setup.

**Reply by VonHeikemen** [reply]
I wanted to do something like that with lsp-zero `v1.x` (two years ago) but it didn't work out. The problems `v1.x` had do not get solve with code (or documentation). There was never a community of lsp-zero users willing to help each other. I could literally build the plugin you want and I think it would fail if that plugin doesn't have a community.

**Reply by pm4rcin** [reply]
I guess it'd require support from Mason maintainers and redirect people here. It would be possible in theory but require strong cooperation.

**Comment by boydkelly** [comment]
Well thanks so much for this amazing project!   As mentioned I am not a programmer, but happens that I have followed lsp since pre neovim 5.  I am just super satisfied with my 23 liner lspzero:

````
vim.g.lsp_zero_extend_cmp = 0
local lsp_zero = require "lsp-zero"
lsp_zero.extend_lspconfig()

lsp_zero.on_attach(function(client, bufnr)
  lsp_zero.default_keymaps { buffer = bufnr }
end)

local icons = require "config.icons"
local lsp_servers = require("config").lsp_servers

lsp_zero.set_sign_icons {
  error = icons.diagnostics.Error,
  warn = icons.diagnostics.Warning,
  hint = icons.diagnostics.Hint,
  info = icons.diagnostics.Information,
}

for _, server in ipairs(lsp_servers) do
  require("lspconfig")[server].setup {}
end

local lua_opts = lsp_zero.nvim_lua_ls()
require("lspconfig").lua_ls.setup(lua_opts)
````
And for my own yaml config i have in after/ftplugin/yaml.lua:
````
local yaml_opts = require "config.lsp.yaml3"
require("lspconfig").yamlls.setup(yaml_opts)
````
I was able to totally remove mason and mason-lspconfig etc.   I'm now seeing that its really just extra fluff.

I now have my own 'mason-tools-installer.sh'  ;)   which provides me all the tools for both nvim and hx.   (It would be nice though to have the auto_update of the 'real' mason-tools-installer...)   And with hx there is really **nothing** else to do.  Just install the tool.   It provides lua with inline diagnostics and stuff I have not seen in neovim yet!!!  Zero config.

````
#!/usr/bin/bash
XCH=${XDG_CACHE_HOME:-$HOME/.cache}

tag_name() {
  latest_release=$(curl -s https://api.github.com/repos/$1/releases/latest)
  tagname=$(echo "$latest_release" | jq -r '.tag_name') && export tagname
  echo "Repo: $1"
  echo "Installing version: $tagname"
}

export N_PREFIX=$HOME/.local/n

if [ ! -f $N_PREFIX/bin/n ]; then
  sed -i /N_PREFIX/d $HOME/.bashrc
  sed -i /N_PREFIX/d $HOME/.bashrc.d/npm.sh
  curl -L https://bit.ly/n-install N_PREFIX=$HOME/.local/n | bash -s -- -y
fi
grep N_PREFIX "$HOME"/.bashrc >>"$HOME"/.bashrc.d/npm.sh && sed -i /N_PREFIX/d "$HOME"/.bashrc
source "$HOME"/.bashrc.d/npm.sh

n latest

packages=(
dockerfile-language-server-nodejs
bash-language-server
yaml-language-server
markdownlint-cli
vscode-langservers-extracted
svelte-language-server
typescript-language-server
eslint-lsp
prettier
)

[[ "$1" == "npm" ]] && printf "%s@latest\n" "${packages[@]}" | xargs npm i -g

repo=google/yamlfmt
tag_name $repo
version=$(echo ${tagname} | sed 's/v//')
download_url="https://github.com/$repo/releases/download/${tagname}/yamlfmt_${version}_Linux_x86_64.tar.gz"
[[ -f ${XCH}/yamlfmt-$tagname.tgz ]] || wget -q "$download_url" -O "${XCH}/yamlfmt-$tagname.tgz"
tar -C "$HOME"/.local/bin/ -zxf "${XCH}/yamlfmt-$tagname.tgz"

repo=redhat-developer/vscode-xml
tag_name $repo
download_url="https://github.com/$repo/releases/download/$tagname/lemminx-linux.zip"
[[ -f ${XCH}/lemminx-linux.zip ]] || wget -q "$download_url" -O "${XCH}/lemminx-linux.zip"
unzip -q -o "${XCH}/lemminx-linux.zip" -d "${XCH}"/  && cp "${XCH}/lemminx-linux" "$HOME/.local/bin/lemminx"

repo=LuaLS/lua-language-server
tag_name $repo
download_url="https://github.com/$repo/releases/download/$tagname/lua-language-server-$tagname-linux-x64.tar.gz"
[[ -f ${XCH}/lua-language-$tagname-linux-x64.tar.gz ]] || wget -q "$download_url" -O ${XCH}/lua-language-$tagname-linux-x64.tar.gz
tar -xf "${XCH}/lua-language-$tagname-linux-x64.tar.gz" -C ~/.local/

repo=JohnnyMorganz/StyLua
tag_name $repo
download_url="https://github.com/$repo/releases/download/$tagname/stylua-linux-x86_64.zip"
[[ -f ${XCH}/stylua-linux-x86_64.zip ]] || wget -q "$download_url" -O ${XCH}/stylua-linux-x86_64.zip
unzip -q -o ${XCH}/stylua-linux-x86_64.zip -d "$HOME/.local/bin/"

repo=artempyanykh/marksman
tag_name $repo
download_url=https://github.com/$repo/releases/download/$tagname/marksman-linux-x64
[[ -f "$HOME"/.local/bin/marksman ]] || wget -q "$download_url" -O "$HOME"/.local/bin/marksman
chmod +x "$HOME"/.local/bin/marksman

repo=Feel-ix-343/markdown-oxide
tag_name $repo
download_url="https://github.com/$repo/releases/download/$tagname/markdown-oxide-$tagname-x86_64-unknown-linux-gnu.tar.gz"
[[ -f "${XCH}/markdown-oxide-$tagname-x86_64-unknown-linux-gnu.tar.gz" ]] || wget -q "$download_url" -O "${XCH}/markdown-oxide-$tagname-x86_64-unknown-linux-gnu.tar.gz"
tar -xvf "${XCH}/markdown-oxide-$tagname-x86_64-unknown-linux-gnu.tar.gz" --strip-components=1 -C "$HOME"/.local/bin/
chmod +x "$HOME"/.local/bin/markdown-oxide

repo=tamasfe/taplo
tag_name $repo
download_url="https://github.com/$repo/releases/download/$tagname/taplo-full-linux-x86_64.gz"
[[ -f "${XCH}/$tagname-taplo-full-linux-x86_84.gz" ]] || wget -q "$download_url" -O "${XCH}/$tagname-taplo-full-linux-x86_64.gz"
gunzip -kc "${XCH}"/$tagname-taplo-full-linux-x86_64.gz > "$HOME"/.local/bin/taplo
chmod +x "$HOME"/.local/bin/taplo

````

**Reply by pm4rcin** [reply]
Would you tell what are the things that are present in hx while not being in nvim?

**Comment by boydkelly** [comment]
I'm really not the expert, but hx feels somewhat uncustomizeable without plugins etc.   But in the realm of lzpzero its great!   There is no mason installer.   So you are on your own to download and/or compile the language servers.   But if the binary is in your path there is _**nothing**_ else to do to have a well functioning lsp+treesitter config.   Run hx --health from the command line and you will see what is supported and active as per below.   I imagine this is all the stuff from lspconfig built in to helix?   

So I could imagine for lspzero a plugin where even running setup is optional.   Just require lspzero and if you have the lsp binary in your path it works when you open the filetype.   And of course you could customize it yourself when necessary.

Off topic:   Anyway I installed helix as the recent versions of nvim-treesitter will not handle my 44000 line yaml file,  so thats what made me try helix and zed as they both use treesitter.  Helix is super snappy, and its really really lsp zero.   I found this highly motiatiing and it caused me to just remove mason all together from my neovim config.   I realize now its very optional.   I use lspzero, a custom yaml and xml config and ny neovim now leans on the same binaries as helix for lsp.   

It made me also think that it would be awesome in the neovim world to have a minimalist mason-installer plugin with no gui type screen.  Just a lua ensure_installed script and plugin to make sure stuff gets updated.   That along with even the current lspzero would be a winner IMO.
````
$ hx --health
Config file: /var/home/bkelly/.config/helix/config.toml
Language file: /var/home/bkelly/.config/helix/languages.toml
Log file: /var/home/bkelly/.cache/helix/helix.log
Runtime directories: /var/home/bkelly/.config/helix/runtime;/var/home/bkelly/.local/helix-24.07-x86_64-linux/runtime
Clipboard provider: wl-paste+wl-copy
System clipboard provider: wl-paste+wl-copy

Language           LSP                DAP                Formatter          Highlight          Textobject         Indent             
====
java               ✘ jdtls            None               None               ✓                  ✓                  ✓                  
javascript         ✓ typescript-la…   ✘                  ✓ prettier         ✓                  ✓                  ✓                  
                   ✓ vscode-eslint…   
jinja              None               None               None               ✓                  ✘                  ✘                  
jsdoc              None               None               None               ✓                  ✘                  ✘                  
json               ✓ vscode-json-l…   None               ✓ prettier         ✓                  ✓                  ✓                  
json5              None               None               None               ✓                  ✘                  ✘                  
jsonc              ✓ vscode-json-l…   None               None               ✓                  ✘                  ✓                  
jsonnet            ✘ jsonnet-langu…   None               None               ✓                  ✘                  ✘                  
jsx                ✓ typescript-la…   None               None               ✓                  ✓                  ✓                  
julia              ✘ julia            None               None               ✓                  ✓                  ✓                  
just               None               None               None               ✓                  ✓                  ✓                  
kdl                None               None               None               ✓                  ✓                  ✓                  
koka               ✘ koka             None               None               ✓                  ✘                  ✓                  
kotlin             ✘ kotlin-langua…   None               None               ✓                  ✘                  ✘                  
latex              ✘ texlab           None               None               ✓                  ✓                  ✘                  
ld                 None               None               None               ✓                  ✘                  ✓                  
ldif               None               None               None               ✓                  ✘                  ✘                  
lean               ✘ lean             None               None               ✓                  ✘                  ✘                  
ledger             None               None               None               ✓                  ✘                  ✘                  
llvm               None               None               None               ✓                  ✓                  ✓                  
llvm-mir           None               None               None               ✓                  ✓                  ✓                  
llvm-mir-yaml      None               None               None               ✓                  ✘                  ✓                  
log                None               None               None               ✓                  ✘                  ✘                  
lpf                None               None               None               ✓                  ✘                  ✘                  
lua                ✓ lua-language-…   None               ✓ stylua           ✓                  ✓                  ✓                  
make               None               None               None               ✓                  ✘                  ✓                  
markdoc            ✘ markdoc-ls       None               None               ✓                  ✘                  ✘                  
markdown           ✓ marksman         None               ✓ prettier         ✓                  ✘                  ✘                  
                   ✓ markdown-oxid…   
````
